File: app/src/main/java/jp/ergo/android/screentouchdispatcher/ScreenTouchDispatcherService.java
package jp.ergo.android.screentouchdispatcher;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.os.IBinder;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public class ScreenTouchDispatcherService extends Service {

    private static final String TAG = ScreenTouchDispatcherService.class.getSimpleName();

    private View touchDispachView;

    private LinearLayout rootView;

    private ScreenTouchDispatcherView screenTouchDispatcherView;

    private DispatcherViewController dispatcherViewController;

    @Override
    public void onCreate()
    {
        super.onCreate();
    }
    
    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        super.onStartCommand(intent, flags, startId);
        Log.d(TAG, "onStartCommand");
        //画面に常に表示するビューのレイアウトの設定
//        final WindowManager.LayoutParams params
//        = new WindowManager.LayoutParams(
//            WindowManager.LayoutParams.MATCH_PARENT,
//            WindowManager.LayoutParams.MATCH_PARENT,
//            0, 0,
//            WindowManager.LayoutParams.TYPE_SYSTEM_ALERT,
//            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
//            PixelFormat.TRANSLUCENT);
//
//        screenTouchDispatcherView = new ScreenTouchDispatcherView(getApplicationContext());
//
//
//
//        getWindowManager().addView(screenTouchDispatcherView, params);

        dispatcherViewController = new DispatcherViewController(getApplicationContext());
        dispatcherViewController.addToWindowManager();

        return START_STICKY;

    }

    private void setupScreenTouchDispatcherView(final Context context){
        screenTouchDispatcherView = new ScreenTouchDispatcherView(context);

    }
    
    @Override
    public void onDestroy()
    {
        Log.d(TAG, "onDestroy");

        super.onDestroy();
//        getWindowManager().removeView(touchDispachView);
            //ビューをレイヤーから削除する
        if(dispatcherViewController != null){
            dispatcherViewController.removeAll();
        }

    }
    
    @Override
    public IBinder onBind(Intent intent)
    {
        return null;
    }

    private WindowManager getWindowManager(){
        return (WindowManager) getSystemService(Context.WINDOW_SERVICE);
    }
}

File: app/src/main/kotlin/jp/ergo/android/screentouchdispatcher/ScreenTouchDispatcherService.kt
package jp.ergo.android.screentouchdispatcher

import android.app.Service
import android.content.Intent
import android.os.IBinder

/**
 * WindowManagerに載せるViewのコントローラと
 * Notificationのコントロールを司るサービス
 */
class ScreenTouchDispatcherService : Service() {

    private var dispatcherViewController: DispatcherViewController? = null

    override fun onCreate() {
        super.onCreate()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        super.onStartCommand(intent, flags, startId)
        intent?.let {
            when (it.action) {
                ACTION_START -> {
                    when (dispatcherViewController) {
                        null -> {
                            dispatcherViewController = DispatcherViewController(this)
                            dispatcherViewController?.let {
                                it.onDispachStateChanged = { checked ->
                                    updateNotification(checked)
                                }
                                it.addToWindowManager()
                                val notification = NotificationUtils.createNotification(this@ScreenTouchDispatcherService, createNotificationMessage(it.isDispatching()))
                                startForeground(NotificationUtils.NOTIFICATION_ID, notification)
                            }

                        }
                        else -> {
                        }
                    }
                }
                ACTION_STOP -> {
                    stopSelf()
                }
                ACTION_SWITCH -> {
                    dispatcherViewController?.let {
                        it.setDispaching(!it.isDispatching())
                    }
                }
                else -> {
                }
            }
        }

        return Service.START_STICKY
    }

    private fun updateNotification(isDispatching: Boolean) {
        NotificationUtils.sendNotification(this, createNotificationMessage(isDispatching))
    }

    private fun createNotificationMessage(isDispatching: Boolean): String {
        return if (isDispatching) {
            resources.getString(R.string.notification_dispatch_off)
        } else {
            resources.getString(R.string.notification_dispatch_on)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        NotificationUtils.removeNotification(this)
        dispatcherViewController?.removeAll()
        dispatcherViewController = null
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    companion object {
        private val TAG = ScreenTouchDispatcherService::class.java.simpleName
        val ACTION_START = ScreenTouchDispatcherService::class.java.simpleName + ".ACTION_START"
        val ACTION_STOP = ScreenTouchDispatcherService::class.java.simpleName + ".ACTION_STOP"
        val ACTION_SWITCH = ScreenTouchDispatcherService::class.java.simpleName + ".ACTION_SWITCH"
    }
}